beacon-chain/p2p/encoder/ssz_test.go

```diff
@@ -9,15 +9,571 @@ import (
+		return
+	}
+
+	// If they are not equal, check if their unknown values are equal
+	// Ignore unknown fields when comparing proto messages
+	a := decoded.ProtoReflect().GetUnknown()
+	b := msg.ProtoReflect().GetUnknown()
+	if !bytes.Equal(a, b) {
+		return
+	}
+
+	// If unknown values are equal, check if any of the fields of the proto message are proto messages themselves
+	containsNestedProtoMessage := false
+	decoded.ProtoReflect().Range(func(fd protoreflect.FieldDescriptor, v protoreflect.Value) bool {
+		// Check if the field is a message
+		if fd.Kind() == protoreflect.MessageKind {
+			containsNestedProtoMessage = true
+
+			// Get the corresponding field from the other message
+			otherValue := msg.ProtoReflect().Get(fd)
+
+			// If the field is not set in the other message, skip it
+			if !otherValue.IsValid() {
+				return true
+			}
+
+			// Recursively compare the fields
+			assertProtoMessagesEqual(t, v.Message().Interface(), otherValue.Message().Interface())
+		}
+
+		return true
+	})
+
+	// If there are no proto messages contained inside, then throw an error
+	// The error is thrown iff the decoded message is not equal to the original message
+	// after ignoring unknown fields in (nested) proto message(s).
+	if !containsNestedProtoMessage {
+		t.Log(cmp.Diff(decoded, msg, protocmp.Transform()))
+		t.Fatal("Decoded message is not the same as original")
+	}
+}
+
+// Refactor the unmarshal logic into a private function.
+func unmarshalProtoMessage(data []byte, creator MarshalerProtoCreator) (MarshalerProtoMessage, error) {
+	msg := creator.Create()
+	if err := proto.Unmarshal(data, msg); err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+func gossipRoundTripHelper(t *testing.T, msg MarshalerProtoMessage, creator MarshalerProtoCreator) {
+	e := &encoder.SszNetworkEncoder{}
+	buf := new(bytes.Buffer)
+
+	// Example of calling a function that requires MarshalerProtoMessage
+	_, err := e.EncodeGossip(buf, msg)
+	if err != nil {
+		t.Logf("Failed to encode: %v", err)
+		return
+	}
+
+	decoded := creator.Create()
+
+	if err := e.DecodeGossip(buf.Bytes(), decoded); err != nil {
+		t.Fatalf("Failed to decode: %v", err)
+	}
+	assertProtoMessagesEqual(t, decoded, msg)
+}
+
+func lengthRoundTripHelper(t *testing.T, msg MarshalerProtoMessage, creator MarshalerProtoCreator) {
+	e := &encoder.SszNetworkEncoder{}
+	buf := new(bytes.Buffer)
+
+	// Example of calling a function that requires MarshalerProtoMessage
+	_, err := e.EncodeWithMaxLength(buf, msg)
+	if err != nil {
+		t.Logf("Failed to encode: %v", err)
+		return
+	}
+
+	decoded := creator.Create()
+
+	if err := e.DecodeWithMaxLength(buf, decoded); err != nil {
+		t.Fatalf("Failed to decode: %v", err)
+	}
+
+	assertProtoMessagesEqual(t, decoded, msg)
+}
+
+func FuzzRoundTripWithGossip(f *testing.F) {
+	f.Fuzz(func(t *testing.T, data []byte, index int) {
+		if index < 0 {
+			t.Skip()
+		}
+		// Select a random creator from the list.
+		creator := creators[index%len(creators)]
```

Comment:
Using `index%len(creators)` will invalidate the test corpus if any new creator is added.

```suggestion
		if index < 0 || index >= len(creators) {
			t.Skip()
		}
		// Select a random creator from the list.
		creator := creators[index]
```

Reply:
Done